rfcs/20250421-source-target-pairs.md

```diff
@@ -0,0 +1,134 @@
+# [RFC] Add source-target pairs to send/recv ops
+
+Status: In Review<br/>
+Initial version: 04/21/2025<br/>
+Last updated: 04/21/2025<br/>
+Discussion thread: N/A
+
+## Overview
+
+This RFC proposes adding a new attribute `source_target_pairs` to `send` and
+`recv` ops. `source_target_pairs` allows users to specify peer-to-peer
+communication patterns using global device IDs (zero-indexed integers).
+Currently this feature is only available on GPUs.
+
+## Background
+
+SPMD-based pipeline parallelism relies on optimizations in XLA to pipeline
+send/recv operations in such a way that compute and communication are
+overlapped. The user expresses this through collective permutes and relies on
+XLA to decompose these into send/recv operations, which are then pipelined
+separately, allowing for the staggering that is unique to pipeline parallelism.
+The limitation of this approach is that it encapsulates the latency hiding
+mechanism in the compiler and allows for little control by the user. When this
+mechanism fails, the user has little choice but to debug XLA itself. This RFC is
+proposed in conjunction with exposing send/recv operations through the JAX
+`shard_map` API.
+
+## Proposed Specification
```

Comment:
Can you add a section for "A Note on Backward Compatibility", denoting that we're technically making the semantics more strict for send given that any instances of `send is_host_transfer=false` that are serialized will no longer be deserializable, but this is likely OK since this would have been undefined behavior as it is.

Reply:
cc @frgossen to confirm this as well